modules/pallets/collator-manager/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+extern crate alloc;
+pub mod currency_adapter;
+
+use polkadot_sdk::*;
+
+use alloc::vec::Vec;
+pub use pallet::*;
+
+/// A trait for providing a list of collator candidates.
+pub trait CandidateProvider<ValidatorId> {
+	/// Returns the current list of collator candidates.
+	fn candidates() -> Vec<ValidatorId>;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::traits::{
+		Currency, Get, fungibles,
+		tokens::{Fortitude, Precision, Preservation},
+	};
+	use pallet_session::SessionManager;
+	use polkadot_sdk::frame_support::traits::fungibles::Mutate;
+	use sp_runtime::traits::Zero;
+	use sp_staking::SessionIndex;
+	use sp_std::vec::Vec;
+
+	type BalanceOf<T> = <<T as Config>::ReputationCurrency as Currency<
+		<T as frame_system::Config>::AccountId,
+	>>::Balance;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: polkadot_sdk::frame_system::Config + pallet_session::Config {
+		type ReputationCurrency: Currency<Self::AccountId>;
+		type CandidateProvider: CandidateProvider<Self::ValidatorId>;
+		#[pallet::constant]
+		type ReputationAssetId: Get<
+			<Self::ReputationAssets as fungibles::Inspect<Self::AccountId>>::AssetId,
+		>;
+		type ReputationAssets: fungibles::Mutate<Self::AccountId, Balance = BalanceOf<Self>>;
+		#[pallet::constant]
+		type DesiredCollators: Get<u32>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		NewCollatorSet(Vec<T::AccountId>),
+		ReputationReset(T::AccountId),
+	}
+
+	impl<T: Config> SessionManager<T::AccountId> for Pallet<T>
+	where
+		T::ValidatorId: Into<T::AccountId> + Clone,
+		T::AccountId: From<T::ValidatorId>,
+		T::AccountId: Into<T::ValidatorId> + Clone,
+		T::ValidatorId: From<T::AccountId>,
+	{
+		fn new_session(_new_index: SessionIndex) -> Option<Vec<T::AccountId>> {
+			let active_collators = <pallet_session::Pallet<T>>::validators();
+			let desired_collators = T::DesiredCollators::get() as usize;
+
+			let mut new_set_validators: Vec<T::ValidatorId> = Vec::new();
+
+			// select from registered candidates who are not in the current active set
+			// with session keys and highes balances.
+			let mut candidates = T::CandidateProvider::candidates();
+			candidates.retain(|c| {
+				!active_collators.contains(c) && pallet_session::NextKeys::<T>::get(&c).is_some()
+			});
+			candidates.sort_by_key(|a| {
+				let account_id: T::AccountId = a.clone().into();
+				T::ReputationCurrency::total_balance(&account_id)
+			});
+			candidates.reverse();
+			new_set_validators.extend(candidates.into_iter().take(desired_collators));
+
+			// fill remaining slots with the best of the previous set.
+			if new_set_validators.len() < desired_collators {
+				let needed = desired_collators - new_set_validators.len();
+				let mut reused_collators = active_collators.clone();
+
+				reused_collators.retain(|c| !new_set_validators.contains(c));
```

Comment:
This line doesn't do anything. The current collators have already been filtered out in line 89